### PR TITLE
Feature/stateless auth

### PR DESCRIPTION
**Pull request by cyleecode**
Authentication middleware to validate /project endpoint using `Authorization` header and `walletAddress` header.

### CODE REVIEW

File: src/middleware/auth-handler.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        // We also check the token expiration is not too far out, which would be a red flag.
+        if (Number(decodedNote[1]) < Date.now() || Number(decodedNote[1]) > Date.now() + day1 + minutes30) {
+            throw new TokenExpiredError()
+        }
+
+        // We check if the params match the ones we set in the front-end
+        if (
+            toCheck.firstRound === 10 &&
+            toCheck.lastRound === 10 &&
+            decodedNote[0] === 'https://testnet.terragrids.org/' &&
```
Comment (lorenzopolidori) [inline src/middleware/auth-handler.js:45]: This should be dependent on the environment:
```js
decodedNote[0] === `https://${env}.terragrids.org/`
```
where
```js
const env = process.env.ENV === 'dev' ? 'testnet' : 'app'
```

**Review by lorenzopolidori — CHANGES_REQUESTED** [review]

File: src/app.js [hunk 2]
```diff
@@ -376,4 +377,4 @@ router.post('/files/upload', bodyparser(), async ctx => {
     ctx.status = 201
 })
 
-app.use(requestLogger).use(errorHandler).use(router.routes()).use(router.allowedMethods())
+app.use(requestLogger).use(errorHandler).use(authHandler).use(router.routes()).use(router.allowedMethods())
```
[the remark this answers is not in the document]
Comment (cyleecode) [reply src/app.js]: Fix with commit https://github.com/terragrids/api/pull/15/commits/4853f8b496b05315fc68ed27f0e1b4f1d83d5e68

File: src/utils/auth-path.js [hunk 3]
```diff
@@ -0,0 +1 @@
+export const authenticatePath = ['/project']
```
[the remark this answers is not in the document]
Comment (cyleecode) [reply src/utils/auth-path.js]: Fix with https://github.com/terragrids/api/pull/15/commits/69e5ffe720f71b17bc8ce8c147afdf3f239f7a17

File: src/middleware/auth-handler.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import { TokenInvalidError } from '../error/token-invalid-error.js'
+import { day1, minutes30 } from '../utils/constants.js'
+import { authenticatePath } from '../utils/auth-path.js'
+import { TokenExpiredError } from '../error/token-expired-error.js'
+
+export default async function authHandler(ctx, next) {
+    await next()
+    //check if path is allow for non auth
+    if (authenticatePath.includes(ctx.request.url)) {
+        const accountId = ctx.headers?.walletaddress
```
[the remark this answers is not in the document]
Comment (cyleecode) [reply src/middleware/auth-handler.js]: Updated with https://github.com/terragrids/api/pull/15/commits/e913019eb5533e887889c15cf2f4219b5596a408

File: src/middleware/auth-handler.js (hunk 1 above)
Comment (cyleecode) [reply src/middleware/auth-handler.js]: Pending to change dapp repo signToken note

File: src/middleware/auth-handler.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        const token = ctx.headers?.authorization?.split(' ')
+        if (token?.length !== 2) throw new TokenInvalidError()
+
+        //converting the base64 encoded tx back to binary data
```
[the remark this answers is not in the document]
Comment (cyleecode) [reply src/middleware/auth-handler.js]: https://github.com/terragrids/api/pull/15/commits/3bd128086ece6f318ef962aec95b5a31fb1b7c5e

**Review by lorenzopolidori — APPROVED** [review]

**Pull request merged**